### PR TITLE
start upgrading V8 to 7.7.299

### PR DESCRIPTION
**Pull request by dothebart**
Starting to deploy the new V8

### CODE REVIEW

**Comment by dothebart** [comment]
@cclaus @refack for the errors obvious to me I've provided pull requests, it would be good if you could lend me a helping hand on identifying my currently remaining issue which doesn't error out, but simply produces errnous behaviour:

I'm using the nodejs gypfiles, and invoke gyp like this:
```
arangosrc = ~/src/arangodb
cd "${arangosrc}/3rdParty/V8/v7.1.302.28" && /usr/bin/python3 "${arangosrc}/3rdParty/V8/v7.1.302.28/tools/gyp/gyp_main.py" --generator-output="${arangosrc}/build/3rdParty/V8/v7.1.302.28" -Dv8_target_arch=x64 -Dv8_host_arch=x64 "-Dbyteorder='little'" -Dv8_enable_i18n_support=1 gypfiles/all.gyp --format make -S.x64.release -Igypfiles/standalone.gypi -Dmac_deployment_target=10.11 --depth=. -Dstandalone_static_library=1 -Dlinux_use_bundled_gold=0 -Dlinux_use_gold_flags=0 -Dv8_no_strict_aliasing=1 -Dwerror= -Dclang=0 -Dhost_clang= -Dv8_embed_script= -Dembed_script= -Dwarmup_script= -Dv8_enable_embedded_builtins=0 -Dembedded_builtins_snapshot_src= -Dv8_enable_handle_zapping=0 -Dv8_use_snapshot=0 -DV8_ROOT="${arangosrc}/3rdParty/V8/v7.1.302.28"
```
which will run through, however it seems that some files are missing partially; If I get that correctly there is a module loading source file lists from .gni files, right? The node-gyp generated makefiles contain dependencies to `src/base/bits.cc`, but doen't add instructions howto build them. So it gyp somehow knows `src/base/bits.cc`, but doesn't generate all neccessary places. Most probably bits is just the first file, more would follow. 

Any hints?

**Comment by dothebart** [comment]
Ok, V8_ROOT has to be relative, if its absolute, everything breaks.

**Comment by ObiWahn** [comment]
https://github.com/arangodb/arangodb/pull/10276 get this as soon as it lands in devel.

**Comment by ObiWahn** [comment]
```
['/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp/pylib', '/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp', '/usr/lib/python37.zip', '/usr/lib/python3.7', '/usr/lib/python3.7/lib-dynload', '/usr/local/lib/python3.7/dist-packages', '/usr/lib/python3/dist-packages', '/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp/..', '/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp/../../gypfiles', '/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp/../../third_party/binutils']
Traceback (most recent call last):
  File "/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp/gyp_main.py", line 53, in <module>
    import gyp
ImportError: bad magic number in 'gyp': b'\x03\xf3\r\n'
make[2]: *** [3rdParty/V8/CMakeFiles/v8_build.dir/build.make:106: 3rdParty/V8/v8_build-prefix/src/v8_build-stamp/v8_build-configure] Error 1
make[1]: *** [CMakeFiles/Makefile2:739: 3rdParty/V8/CMakeFiles/v8_build.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

ImportError: bad magic number in 'gyp': b'\x03\xf3\r\n'

Can be fixed by running `find . -name "*.pyc" -delete` in source.

**Comment by ObiWahn** [comment]
Traceback (most recent call last):
  File "/home/xxx/checkouts/arangodb5/3rdParty/V8/v7.1.302.28/tools/gyp/gyp_main.py", line 56, in <module>
    sys.exit(gyp.script_main())
AttributeError: module 'gyp' has no attribute 'script_main'

**Comment by ObiWahn** [comment]
@dothebart do you know which gyp is loaded and if the version matches? I had no luck with `gyp.__file__` or the `inspect` module locating gyp.

**Comment by cclauss** [comment]
> Ok, V8_ROOT has to be relative, if its absolute, everything breaks.

Does the directive `from __future__ import absolute_import` help or hurt?

https://docs.python.org/3.8/library/__future__.html

File: lib/V8/v8-globals.h [hunk 1]
```diff
@@ -756,7 +760,7 @@ TRI_v8_global_t* TRI_GetV8Globals(v8::Isolate*);
 template <typename TARGET>
 bool TRI_V8_AddProtoMethod(v8::Isolate* isolate, TARGET tpl, v8::Handle<v8::String> name,
                            v8::FunctionCallback callback, bool isHidden = false) {
-  // hidden method
+  /*  // hidden method
```
Comment (ObiWahn) [inline lib/V8/v8-globals.h]:
```
-771,7 +771,7 @@ bool TRI_V8_AddProtoMethod(v8::Isolate* isolate, TARGET tpl, v8::Handle<v8::Stri
 
 /// @brief adds a method to an object
 template <typename TARGET>
-inline bool TRI_V8_AddMethod(v8::Isolate* isolate, TARGET tpl, v8::Handle<v8::String> name,
+inline bool TRI_V8_AddMethod(v8::Isolate* isolate, v8::MaybeLocal<TARGET> tpl, v8::Handle<v8::String> name,
                              v8::Handle<v8::FunctionTemplate> callback,
                              bool isHidden = false) {
   // hidden method
@@ -787,8 +787,9 @@ inline bool TRI_V8_AddMethod(v8::Isolate* isolate, TARGET tpl, v8::Handle<v8::St
 }
 
 template <typename TARGET>
-inline bool TRI_V8_AddMethod(v8::Isolate* isolate, TARGET tpl, v8::Handle<v8::String> name,
+inline bool TRI_V8_AddMethod(v8::Isolate* isolate, v8::MaybeLocal<TARGET> tpl, v8::Handle<v8::String> name,
                              v8::FunctionCallback callback, bool isHidden = false) {
+
```
maybe we should do this instead

**Comment by ObiWahn** [comment]
Can you split the PR. It becomes impossible do to good review here.

- 1st pr (this) -  copy the files from upstream
- 2nd pr that has this pr as base - modifications in arangodb

File: lib/V8/v8-globals.h (hunk 1 above)
Comment (ObiWahn) [reply lib/V8/v8-globals.h]: ping

**Comment by ObiWahn** [comment]
Why do we use `FromMaybe()` and not `ToLocalChecked()` is there a _reason_ why we always use _default values_?

**Comment by ObiWahn** [comment]
```
@@ -275,7 +275,7 @@ bool AddFollower::start(bool&) {
                        trx.add(VPackValue(planPath));
                        {
                          VPackArrayBuilder serverList(&trx);
-                         for (VPackSlice srv : VPackArrayIterator(plan)) {
+                         for (auto const& srv : VPackArrayIterator(plan)) {
                            trx.add(srv);
                          }
                          for (auto const& srv : chosen) {
```

Please do not do such beatification changes that improve nothing and just destroy type information. Not every loop has to look the same. You are allowed to reason! @dothebart Is a `VpackSlice` bigger than a pointer?

**Comment by dothebart** [comment]
Superseeded by https://github.com/arangodb/arangodb/pull/10978 - which will directly go to the currently available V8 version.

**Pull request closed without merging**